File: 85-maximal-rectangle/maximal-rectangle.java
import java.util.*;

class Solution {

    public int maximalRectangle(char[][] matrix) {

        if (matrix == null || matrix.length == 0 || matrix[0].length == 0)
            return 0;

        int rows = matrix.length;
        int cols = matrix[0].length;

        int[] arr = new int[cols];
        int max = 0;

        for (int i = 0; i < rows; i++) {

            // Build histogram
            for (int j = 0; j < cols; j++) {
                if (matrix[i][j] == '1')
                    arr[j]++;
                else
                    arr[j] = 0;
            }

            max = Math.max(max, largestRectangle(arr));
        }

        return max;
    }

    public int largestRectangle(int[] arr) {

        int n = arr.length;

        int[] nse = new int[n];
        int[] pse = new int[n];

        Arrays.fill(nse, n);
        Arrays.fill(pse, -1);

        ArrayDeque<Integer> st = new ArrayDeque<>();

        // Next Smaller Element
        st.push(n - 1);

        for (int i = n - 2; i >= 0; i--) {
            while (!st.isEmpty() && arr[st.peek()] >= arr[i]) {
                st.pop();
            }

            if (!st.isEmpty()) {
                nse[i] = st.peek();
            }

            st.push(i);
        }

        st.clear();

        // Previous Smaller Element
        st.push(0);

        for (int i = 1; i < n; i++) {
            while (!st.isEmpty() && arr[st.peek()] >= arr[i]) {
                st.pop();
            }

            if (!st.isEmpty()) {
                pse[i] = st.peek();
            }

            st.push(i);
        }

        int max = 0;

        for (int i = 0; i < n; i++) {
            int width = nse[i] - pse[i] - 1;
            int area = arr[i] * width;
            max = Math.max(max, area);
        }

        return max;
    }
}

File: 85-maximal-rectangle/maximal-rectangle.cpp
class Solution {
public:

    vector<int> NSE(vector<int>& arr){

        int n=arr.size();

        vector<int> nse(n);

        stack<int> st;

        for(int i=n-1;i>=0;i--){

            while(!st.empty() && arr[st.top()]>=arr[i])
                st.pop();

            if(st.empty())
                nse[i]=n;
            else
                nse[i]=st.top();

            st.push(i);
        }

        return nse;
    }

    vector<int> PSEE(vector<int>& arr){

        int n=arr.size();

        vector<int> pse(n);

        stack<int> st;

        for(int i=0;i<n;i++){

            while(!st.empty() && arr[st.top()]>=arr[i])
                st.pop();

            if(st.empty())
                pse[i]=-1;
            else
                pse[i]=st.top();

            st.push(i);
        }

        return pse;
    }

    int largestHistogram(vector<int>& height){

        vector<int> nse=NSE(height);

        vector<int> pse=PSEE(height);

        int ans=0;

        for(int i=0;i<height.size();i++){

            int width=nse[i]-pse[i]-1;

            ans=max(ans,width*height[i]);
        }

        return ans;
    }

    int maximalRectangle(vector<vector<char>>& matrix) {

        if(matrix.empty()) return 0;

        int n=matrix.size();

        int m=matrix[0].size();

        vector<int> height(m,0);

        int ans=0;

        for(int i=0;i<n;i++){

            for(int j=0;j<m;j++){

                if(matrix[i][j]=='1')
                    height[j]++;
                else
                    height[j]=0;
            }

            ans=max(ans,largestHistogram(height));
        }

        return ans;
    }
};